### PR TITLE
Add Decimal128

### PR DESCRIPTION
**Pull request by nielsenko**
Add support for `Decimal128`. This behaves as you would expect from IEEE754, including behaviour of NaN comparison, ie. `Nan == NaN` is `false`. Note that you can still query for a `NaN` value in the DB as realm-core treats `NaN` as any other (single) value, to have a total order.

Be aware that the Dart `compareTo` method is implemented to honor the total ordering of IEEE754-2008 paragraf 5.10, which mimic the behaviour of Dart `double`. Note that this means that `compareTo` will order any values, including `+/- NaN`, `+/- Infinity` as well as `+/- 0`. This implies the `compareTo` operation is different from the partial ordering given by `operator==`, `operator<`, and `operator>`. For example, IEEE doubles impose that `Decimal128.zero == -Decimal128.zero` and all comparison operations on NaN return false.

This function imposes a complete ordering for doubles. When using `compareTo`, the following properties hold:

- All NaN values are considered equal, and greater than any numeric value.
- `-Decimal128.zero` is less than `Decimal128.zero` (and the integer 0), but greater than any non-zero negative value.
- Negative infinity is less than all other values and positive infinity is greater than all non-NaN values.
- All other values are compared using their numeric value.

### CODE REVIEW

**Comment by dotjon0** [comment]
so pleased this is in motion @nielsenko , thank you soooo much!

**Comment by nielsenko** [comment]
@dotjon0 I will need to postpone this work about a week, but then I'll return to it. I hope it is not inconvenient.

Note that it is a fairly rudimentary implementation supporting just basic the arithmetic operations (`+`,`-`,`*`,`/`, and `abs()`)and the ability to convert to from `int`, `double`, and `String`.

**Comment by dotjon0** [comment]
> @dotjon0 I will need to postpone this work about a week, but then I'll return to it. I hope it is not inconvenient.
> 
> Note that it is a fairly rudimentary implementation supporting just basic the arithmetic operations (`+`,`-`,`*`,`/`, and `abs()`)and the ability to convert to from `int`, `double`, and `String`. 
> 
> 

@nielsenko no need to apologise, we are just very grateful for this feature going in! Functionality sounds great. Thank you so much

**Review by desistefanova — COMMENTED** [review]
I think it will be good to have tests about performing math operation between different types. For example:
decimal128 * int, decimal128 + double, decimal128 * long ...

File: lib/src/native/realm_core.dart [hunk 1]
```diff
@@ -47,9 +47,26 @@ import '../user.dart';
 import '../set.dart';
 import 'realm_bindings.dart';
 
-late RealmLibrary _realmLib;
+const bugInTheSdkMessage = "This is likely a bug in the Realm SDK - please file an issue at https://github.com/realm/realm-dart/issues";
 
-final _RealmCore realmCore = _RealmCore();
+final _realmLib = () {
```
Comment (desistefanova) [inline lib/src/native/realm_core.dart:52]: Is this change related to decimal128?

File: lib/src/native/realm_core.dart [hunk 2]
```diff
@@ -47,9 +47,26 @@ import '../user.dart';
...
+    final additionMessage =
+        isFlutterPlatform ? bugInTheSdkMessage : "Did you forget to run `dart run realm_dart install` after upgrading the realm_dart package?";
+    throw RealmException('Realm SDK package version does not match the native library version ($libraryVersion != $nativeLibraryVersion). $additionMessage');
+  }
+  return result;
+}();
+
+// stamped into the library by the build system (see prepare-release.yml)
+const libraryVersion = '1.0.3';
+late String nativeLibraryVersion;
```
Comment (desistefanova) [inline lib/src/native/realm_core.dart:65]: Where the `nativeLibraryVersion` is set?

Comment (nielsenko) [reply lib/src/native/realm_core.dart:65]: line 54

File: lib/src/native/realm_core.dart (hunk 1 above)
Comment (nielsenko) [reply lib/src/native/realm_core.dart:52]: Yes. Allows `lib` to be used before `_RealmCore` is instantiated.

File: lib/src/decimal128.dart [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+import 'dart:convert';
+import 'dart:ffi';
+
+import 'package:ffi/ffi.dart';
+
+import 'native/realm_bindings.dart';
+import 'native/realm_core.dart';
+
+/// A 128-bit decimal floating point number.
+class Decimal128 extends Comparable<Decimal128> {
```
Comment (desistefanova) [inline lib/src/decimal128.dart:28]: Should be exported in realm_class.dart. It is not visible in the API documentation.

File: lib/src/decimal128.dart [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+  /// The value 0.
+  static final zero = Decimal128.fromInt(0);
+
+  /// The value 1.
+  static final one = Decimal128.fromInt(1);
+
+  /// The value 10.
+  static final ten = Decimal128.fromInt(10);
```
Comment (desistefanova) [inline lib/src/decimal128.dart:36]: Wondering, why 1 and 10 are special and have them as properties.

File: lib/src/decimal128.dart [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+  }
+
+  /// Converts a `int` into a [Decimal128].
+  factory Decimal128.fromInt(int value) {
+    return Decimal128._(lib.realm_dart_decimal128_from_int64(value));
+  }
+
+  /// Converts a `double` into a [Decimal128].
+  factory Decimal128.fromDouble(double value) {
+    return Decimal128.parse(value.toString()); // TODO(kn): Find a way to optimize this
```
Comment (desistefanova) [inline lib/src/decimal128.dart:74]: One TODO comment has left here.

File: test/decimal128_test.dart [hunk 6]
```diff
@@ -0,0 +1,305 @@
...
+
+    expect(Decimal128.one.compareTo(Decimal128.nan), -1);
+    expect(Decimal128.nan.compareTo(Decimal128.nan), 0);
+    expect(Decimal128.nan.compareTo(Decimal128.one), 1);
+  });
+
+  repeatTest('Decimal128.compareTo + <, <=, ==, !=, >=, >', (x, xInt, y, yInt) {
+    expect(x.compareTo(y), -(y.compareTo(x)));
+    expect(x.compareTo(y), xInt.compareTo(yInt));
+    expect(x == y, y == x);
```
Comment (desistefanova) [inline test/decimal128_test.dart:243]: expect(x != y, xInt != yInt);

File: test/decimal128_test.dart [hunk 7]
```diff
@@ -0,0 +1,305 @@
...
+    expect(x < y, xInt < yInt);
+    expect(x <= y, y >= x);
+    expect(x <= y, xInt <= yInt);
+    expect(x > y, y < x);
+    expect(x > y, xInt > yInt);
+    expect(x >= y, y <= x);
+    expect(x >= y, xInt >= yInt);
+
+    expect(x.compareTo(x), 0);
+    expect(x == x, isTrue);
```
Comment (desistefanova) [inline test/decimal128_test.dart:255]:
```suggestion
     expect(x != x, isFalse);
     expect(x == x, isTrue);
```

File: test/decimal128_test.dart (hunk 6 above)
Comment (desistefanova) [inline test/decimal128_test.dart:243]:
```suggestion
    expect(x != y, xInt != yInt);
    expect(x == y, y == x);
```

File: test/decimal128_test.dart [hunk 8]
```diff
@@ -0,0 +1,305 @@
...
+
+        body(x, xInt, y, yInt);
+      },
+      times,
+    );
+  });
+}
+
+Future<void> main([List<String>? args]) async {
+  test('Decimal128.nan', () {
```
Comment (desistefanova) [inline test/decimal128_test.dart:53]: What about adding some test for query realm by Decimal128?

File: lib/src/decimal128.dart (hunk 4 above)
Comment (nielsenko) [reply lib/src/decimal128.dart:36]: I expect 0,1 and 10 to be the most used constants, so I const construct them. Obviously the user is free not to use them, or define const constructed values themself. Problem is the Decimal.fromInt cannot be a const constructor, so it is more memory efficient to use `Decimal128.one` vs `Decimal.fromInt(1)` repeatedly.. and also ever so slightly faster

File: lib/src/decimal128.dart (hunk 5 above)
Comment (nielsenko) [reply lib/src/decimal128.dart:74]: Yeah.. it is a bit of a bummer.. Let me look it over again

File: lib/src/native/realm_core.dart (hunk 1 above)
Comment (blagoev) [reply lib/src/native/realm_core.dart:52]: I would suggest we move decimal128.dart into the native directory and keeping the initialization of realmCore as is. 
_realmLib is intentionally hidden from the rest of the SDK. We could declare deciam128 as a partial file to realm_core.dart if we want to keep it separate from the rest of the native invocations. Which I agree is better.

File: lib/src/native/realm_core.dart [hunk 9]
```diff
@@ -47,9 +47,26 @@ import '../user.dart';
...
+  if (libraryVersion != nativeLibraryVersion) {
+    final additionMessage =
+        isFlutterPlatform ? bugInTheSdkMessage : "Did you forget to run `dart run realm_dart install` after upgrading the realm_dart package?";
+    throw RealmException('Realm SDK package version does not match the native library version ($libraryVersion != $nativeLibraryVersion). $additionMessage');
+  }
+  return result;
+}();
+
+// stamped into the library by the build system (see prepare-release.yml)
+const libraryVersion = '1.0.3';
```
Comment (blagoev) [inline lib/src/native/realm_core.dart:64]: This won't work with our prepare-release.yaml. We are looking for a specific string to replace there.

**Review by blagoev — CHANGES_REQUESTED** [review]
I don't see the support for `Decimal128` in Realm models. I think we should expand this PR and add it here.

File: lib/src/decimal128.dart (hunk 4 above)
Comment (blagoev) [reply lib/src/decimal128.dart:36]: Why do we expect these to be frequently used?

File: lib/src/native/realm_core.dart (hunk 1 above)
Comment (blagoev) [inline lib/src/native/realm_core.dart:52]: I would suggest we move decimal128.dart into the native directory and keeping the initialization of realmCore as is.
_realmLib is intentionally hidden from the rest of the SDK. We could declare deciam128 as a partial file to realm_core.dart if we want to keep it separate from the rest of the native invocations. Which I agree is better.

**Comment by nielsenko** [comment]
> I don't see the support for `Decimal128` in Realm models. I think we should expand this PR and add it here.

I have extended the generator to allow `Decimal128` props, and added a simple test

File: lib/src/native/realm_core.dart (hunk 9 above)
Comment (nielsenko) [reply lib/src/native/realm_core.dart:64]: Hmm.. I thought I had updated `prepare-release.yml` let me check..

File: test/dynamic_realm_test.dart [hunk 10]
```diff
@@ -87,16 +87,17 @@ Future<void> main([List<String>? args]) async {
...
       nullableDateProp: date,
       nullableDoubleProp: -123.456,
       nullableObjectIdProp: objectId,
       nullableUuidProp: uuid,
-      nullableIntProp: 123);
+      nullableIntProp: 123,
+      nullableDecimalProp: Decimal128.fromDouble(4242));
 
-  AllTypes _getEmptyAllTypes() => AllTypes('', false, DateTime(0).toUtc(), 0, objectId, uuid, 0);
+  AllTypes _getEmptyAllTypes() => AllTypes('', false, DateTime(0).toUtc(), 0, objectId, uuid, 0, Decimal128.zero);
```
Comment (desistefanova) [inline test/dynamic_realm_test.dart:100]: Decimal128 has to be added also in `_validateDynamic` and `_validateDynamicLists` below.

**Review by desistefanova — COMMENTED** [review]
Will be good to add Decimal128 also to `realm_set_test.dart` and `realm_value_test.dart` tests.

File: lib/src/decimal128.dart (hunk 4 above)
Comment (nielsenko) [reply lib/src/decimal128.dart:37]: Being the additive identity, multiplicative identity, and the base (0, 1, and 10). It also matches https://pub.dev/documentation/decimal/latest/decimal/Decimal-class.html

**Comment by nielsenko** [comment]
> I don't see the support for `Decimal128` in Realm models. I think we should expand this PR and add it here.

I have included that in this PR as well.

**Comment by nielsenko** [comment]
> Will be good to add Decimal128 also to `realm_set_test.dart` and `realm_value_test.dart` tests.

Done. This actually exposed the fact that I forgot to enable `Decimal128` in `RealmValue`

**Comment by nielsenko** [comment]
> I don't see the support for `Decimal128` in Realm models. I think we should expand this PR and add it here.

We now have full support, including usage in mixed.

File: lib/src/native/realm_core.dart (hunk 1 above)
Comment (nielsenko) [reply lib/src/native/realm_core.dart:54]: I tried doing this, but it gets ugly quickly. Fx. we then need to `show` a class from `native/realm_core.dart`. I'm not sure we want to go that way. What do you think?

Comment (nielsenko) [reply lib/src/native/realm_core.dart:55]: You can see the changes in the [Move decimal128.dart to native as parts file of realm_core.dart](https://github.com/realm/realm-dart/pull/1192/commits/c9433333669cbf74c8298df66b136c2c9be2e2c0) commit

File: lib/src/native/realm_core.dart (hunk 9 above)
Comment (nielsenko) [reply lib/src/native/realm_core.dart:67]: Fixed

File: lib/src/decimal128.dart (hunk 5 above)
Comment (nielsenko) [reply lib/src/decimal128.dart]: We got to live with this for now :-/

File: lib/src/native/realm_core.dart (hunk 1 above)
Comment (blagoev) [reply lib/src/native/realm_core.dart:55]: I think its better. I like that we are not importing ffi and bindings stuff into non native part of the SDK.

The only "downside" is we need to do 
> export 'native/realm_core.dart' show Decimal128;

but that's not an issue since we are in control there.

File: test/decimal128_test.dart [hunk 11]
```diff
@@ -0,0 +1,305 @@
...
+        final y = Decimal128.fromInt(yInt);
+
+        body(x, xInt, y, yInt);
+      },
+      times,
+    );
+  });
+}
+
+Future<void> main([List<String>? args]) async {
```
Comment (blagoev) [inline test/decimal128_test.dart:52]: we should import test.dart and `await setupTests` in all tests files. Thus at least keep consistency on init between the test files

File: test/decimal128_test.dart [hunk 12]
```diff
@@ -0,0 +1,305 @@
...
+const int defaultTimes = 100000;
+
+void repeat(dynamic Function() body, [int times = defaultTimes]) {
+  for (var i = 0; i < times; ++i) {
+    body();
+  }
+}
+
+@isTest
+void repeatTest(String description, dynamic Function(Decimal128 x, int xInt, Decimal128 y, int yInt) body, [int times = defaultTimes]) {
```
Comment (blagoev) [inline test/decimal128_test.dart:34]: I can't invoke single test from command line or vs code. Could we make this work

File: lib/src/native/realm_core.dart [hunk 13]
```diff
@@ -47,9 +48,26 @@ import '../user.dart';
...
+late String nativeLibraryVersion;
```
Comment (blagoev) [inline lib/src/native/realm_core.dart:68]: It seems we don't need this field. So 
>nativeLibraryVersion = result.realm_dart_library_version().cast<Utf8>().toDartString();
could be a local variable.

**Review by blagoev — CHANGES_REQUESTED** [review]
All in all its good. I have some suggestions for improvements.

File: test/decimal128_test.dart (hunk 11 above)
Comment (nielsenko) [reply test/decimal128_test.dart:52]: We don't use realms at all in `decimal128_test.dart` so I really have no reason for that. The integration of `Decimal128` with realm is tested elsewhere, like in `realm_object_test.dart`, `realm_value_test.dart`, etc.
But sure..

File: test/decimal128_test.dart (hunk 12 above)
Comment (blagoev) [reply test/decimal128_test.dart:34]: I mean I tried that in VS code. Is it working for you?

File: test/decimal128_test.dart (hunk 11 above)
Comment (blagoev) [reply test/decimal128_test.dart:52]: I see your point, but I am afraid we might miss some initialization we rely on if we diverge here. For example the individual test timeout gets set up,  `test` method gets replace etc. For example I was surprised with my tests timing out while debugging these tests.

File: test/decimal128_test.dart (hunk 12 above)
Comment (blagoev) [reply test/decimal128_test.dart:34]: I just tried and I get the same error 
> No tests match regular expression "^Decimal128 divide( \(variant: .*\))?$".
and I think I have the latest versions of the VSCode extensions

Comment (nielsenko) [reply test/decimal128_test.dart:34]: Yes. How do you invoke it?

Comment (blagoev) [reply test/decimal128_test.dart:34]: Also if I do 
dart test -n "device" -r expanded in the repo the `decimal` tests are not executed

File: lib/src/native/realm_core.dart (hunk 13 above)
Comment (blagoev) [reply lib/src/native/realm_core.dart:68]: Also If I do `dart test -n "devide" -r expanded` in the repo the `decimal` tests are not included in the run at all.

File: test/decimal128_test.dart [hunk 14]
```diff
@@ -0,0 +1,305 @@
...
+  final r = Random(42); // use a fixed seed to make tests deterministic
```
[the remark this answers is not in the document]
Comment (nielsenko) [reply test/decimal128_test.dart]: Fixed, but for the record...

`r` is declared and used twice over just 7 lines.. this is **not** a readability issue. I will fix this just because, but please don't vilify all short variable names. They put emphasis on the _operations_ as opposed to the _variables_.

Or in the words of Bob Nyström (one of the Dart language designers): https://journal.stuffwithstuff.com/2016/06/16/long-names-are-long/

File: test/decimal128_test.dart (hunk 12 above)
Comment (nielsenko) [reply test/decimal128_test.dart:36]: Fixed

File: test/decimal128_test.dart [hunk 15]
```diff
@@ -0,0 +1,305 @@
...
+
+    expect(x.compareTo(x), 0);
+    expect(x == x, isTrue);
+    expect(x < x, isFalse);
+    expect(x <= x, isTrue);
+    expect(x > x, isFalse);
+    expect(x >= x, isTrue);
+  });
+
+  repeatTest('Decimal128.toInt/fromInt roundtrip', (x, xInt, y, yInt) {
```
[the remark this answers is not in the document]
Comment (nielsenko) [reply test/decimal128_test.dart:266]: Okay. This the reg exp that VSCode builds to select the test when you press the little `run` button:
image
I have fixed the test naming

File: src/realm_dart.cpp [hunk 16]
```diff
@@ -53,9 +55,12 @@ RLM_API void realm_dart_initializeDartApiDL(void* data) {
     Dart_InitializeApiDL(data);
 }
 
+using namespace realm;
```
[the remark this answers is not in the document]
Comment (nielsenko) [reply src/realm_dart.cpp:56]: Done

File: lib/src/native/realm_core.dart (hunk 13 above)
Comment (nielsenko) [reply lib/src/native/realm_core.dart]: Try `dart test -n 'divide' -r expanded` instead.

File: src/realm_dart.cpp [hunk 17]
```diff
@@ -53,9 +53,12 @@ RLM_API void realm_dart_initializeDartApiDL(void* data) {
...
+using namespace realm::c_api;
```
Comment (blagoev) [inline src/realm_dart.cpp:57]: Do we need these statements now in this fail?

**Review by blagoev — APPROVED** [review]

File: src/realm_dart.cpp (hunk 17 above)
Comment (nielsenko) [reply src/realm_dart.cpp:57]: No, will remove

File: test/decimal128_test.dart (hunk 8 above)
Comment (nielsenko) [reply test/decimal128_test.dart:57]: Done

**Review by desistefanova — APPROVED** [review]

File: test/decimal128_test.dart (hunk 7 above)
Comment (nielsenko) [reply test/decimal128_test.dart:257]: It is tested in line 261
